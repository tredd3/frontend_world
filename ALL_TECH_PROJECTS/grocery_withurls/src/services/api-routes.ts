const baseUrl = process.env.REACT_APP_BASE_URL || '';
const baseUrlLogin = process.env.REACT_APP_BASE_URL_LOGIN || '';

export const apiRoutes = {
  getNewAccessToken: `${baseUrl}/jm/auth/oauth/v2/token`,
  getDashboard: `${baseUrl}/jiokart/v1/jiogrocery/get/dashboard`,
  viewAllDashboard: `${baseUrl}/jiokart/v1/jiogrocery/view/all/dashboard`,
  getStores: `${baseUrl}/jiokart/v1/get/stores`,
  getCart: `${baseUrl}/jiokart/v1/jiogrocery/get/cart`,
  addCartItem: `${baseUrl}/jiokart/v1/add/cart/item`,
  removeCartItem: `${baseUrl}/jiokart/v1/remove/cart/item`,
  moveCartItem: `${baseUrl}/jiokart/v1/jiogrocery/move/item`,
  getCategories: `${baseUrl}/jiokart/v1/get/categories`,
  getProductList: `${baseUrl}/jiokart/v1/jiogrocery/store/product`,
  checkServiceability: `${baseUrl}/jiokart/v1/jiogrocery/pincode/check`,
  getProductSearch: `${baseUrl}/jiokart/v1/jiogrocery/product/search`,
  getSuggestions: `${baseUrl}/jiokart/v1/jiogrocery/suggestion`,
  addUser: `${baseUrl}/jiokart/v1/jiogrocery/add/user`,
  saveAddress: `${baseUrl}/jiokart/v1/jiogrocery/save/user/address`,
  getAddress: `${baseUrl}/jiokart/v1/jiogrocery/get/user/address`,
  deleteAddress: `${baseUrl}/jiokart/v1/jiogrocery/remove/user/address`,
  getProductDetails: `${baseUrl}/jiokart/v1/jiogrocery/store/product`,
  getSimilarItems: `${baseUrl}/jiokart/v1/jiogrocery/store/similar/items`,
  getOrderDetails: `${baseUrl}/jiokart/v1/jiogrocery/order/details`,
  getOrderHistory: `${baseUrl}/jiokart/v1/jiogrocery/order/listing`,
  getDeliveryModes: `${baseUrl}/jiokart/v1/get/delivery/modes`,
  getLocation: `${baseUrl}/jiokart/v1/jiogrocery/google/places`,
  searchOrderHistory: `${baseUrl}/jiokart/v1/jiogrocery/order/item/search`,
  checkout: `${baseUrl}/jiokart/v1/order/checkout`,
  placeOrder: `${baseUrl}/jiokart/v1/place/order`,
  getUserDetail: `${baseUrl}/jiokart/v1/jiogrocery/get/user`,
  getWishListData: `${baseUrl}/jiokart/v1/jiogrocery/get/item/list`,
  addItemToWishList: `${baseUrl}/jiokart/v1/jiogrocery/add/item`,
  removeItemFromWishList: `${baseUrl}/jiokart/v1/jiogrocery/remove/item`,
  sendOtp: `${baseUrlLogin}/idam/trialwifi/v3/dip/otp/send`,
  verifyOtp: `${baseUrlLogin}/idam/trialwifi/v3/dip/otp/verify`,
  validateuser: `${baseUrl}/cr/v1/validateuser`,
  cancelOrder: `${baseUrl}/jiokart/v1/order/cancel`,
  config: `${baseUrl}/jiokart/v1/logged/customer/configurations`,
  updateUser: `${baseUrl}/jiokart/v1/jiogrocery/update/user`,
  orderRating: `${baseUrl}/jiokart/v1/order/rating`,
  getPaymentRequest: `${baseUrl}/jiokart/v1/get/payment/request`,
  getPaymentResponse: `${baseUrl}/jiokart/v1/get/payment/response`,
  getNotifications: `${baseUrl}/jiokart/v1/jiogrocery/notification/get`,
  saveNotification: `${baseUrl}/jiokart/v1/jiogrocery/notification/save`,
  clearNotifications: `${baseUrl}/jiokart/v1/jiogrocery/clear/notification`,
  removeCartSavedItem: `${baseUrl}/jiokart/v1/jiogrocery/remove/item`
} as const;

const endpointsWithoutAuth = [
  apiRoutes.checkServiceability,
  apiRoutes.getStores,
  apiRoutes.getProductDetails,
  apiRoutes.getSimilarItems,
  apiRoutes.getLocation,
  apiRoutes.sendOtp,
  apiRoutes.getDashboard
];

const endpointsNeedingEnvelope = [
  apiRoutes.getCategories,
  apiRoutes.addCartItem,
  apiRoutes.removeCartItem,
  apiRoutes.getStores,
  apiRoutes.getDeliveryModes,
  apiRoutes.checkout,
  apiRoutes.placeOrder,
  apiRoutes.getPaymentRequest,
  apiRoutes.getPaymentResponse,
  apiRoutes.orderRating
];

const endpointsNeedingStoreId = [
  apiRoutes.getCart,
  apiRoutes.addCartItem,
  apiRoutes.removeCartItem,
  apiRoutes.placeOrder,
  apiRoutes.getDeliveryModes,
  apiRoutes.checkout,
  apiRoutes.placeOrder,
  apiRoutes.saveAddress,
  apiRoutes.getAddress,
  apiRoutes.getPaymentResponse,
  apiRoutes.removeCartSavedItem
];

export const needsEnvelope = (url: string) => endpointsNeedingEnvelope.includes(url);
export const needsAuthentication = (url: string) => !endpointsWithoutAuth.includes(url);
export const needsStoreId = (url: string) => endpointsNeedingStoreId.includes(url);

export const v2ApiRoutes = [
  apiRoutes.getPaymentResponse,
  apiRoutes.getOrderHistory,
  apiRoutes.getOrderDetails,
  apiRoutes.getCart,
  apiRoutes.getDeliveryModes
];
