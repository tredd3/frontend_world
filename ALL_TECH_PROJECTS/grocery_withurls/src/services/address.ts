import promiseMemoize from 'promise-memoize';
import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { getUser, updateUser } from './user';
import { APIBaseAddress, APIAddress, AddressPivot } from '../types';
import { pivotAPIAddress } from './helpers/pivots';

export type APIAddressRequest = APIBaseAddress & {
  addressId: number;
  addedDate: string;
};

// eslint-disable-next-line camelcase, @typescript-eslint/camelcase
export const DEPRECATED_convertAddressToAPIAddress = ({ id, addedDate, ...address }: APIAddress): APIAddressRequest => (
  { addressId: id, addedDate: addedDate.toDateString(), ...address }
);

type AreaType = { pincode: number};
export const isAreaServiceable = promiseMemoize(async ({ pincode }: AreaType) => {
  try {
    await makeApiCall<AreaType, {}>(apiRoutes.checkServiceability, { body: { pincode } });
    return true;
  } catch (e) {
    return false;
  }
}, { resolve: 'json' });

export const getAddresses = promiseMemoize(async (): Promise<AddressPivot[]> => {
  const { userId } = await getUser();
  const { addresses } = await makeApiCall<{ userId: number }, {addresses: APIAddressRequest[]}>(apiRoutes.getAddress, { body: { userId } });
  return addresses.map(pivotAPIAddress);
});

export const getAddressById = async (id: number): Promise<AddressPivot | undefined> => {
  const addresses = (await getAddresses()) || [];
  return addresses.find(address => address.id === id);
};

type UpdateAddressRequest = (APIAddressRequest | APIBaseAddress) & { userId?: number };

export const updateAddress = async (address: Omit<APIBaseAddress, 'address'> | APIAddress): Promise<APIAddress> => {
  const { id } = address as APIAddress;
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { addedDate, ...remainingAddress } = address as APIAddress;
  const newAddress = await makeApiCall<UpdateAddressRequest, APIAddressRequest>(apiRoutes.saveAddress, {
    body: {
      ...(id && { addressId: id }),
      ...({ userId: (await getUser()).userId }),
      ...remainingAddress,
      address: `${remainingAddress.cityName}, ${remainingAddress.pincode}`
    }
  });

  getAddresses.clear();
  return pivotAPIAddress(newAddress);
};

export const addAddress = (address: Omit<APIBaseAddress, 'address'>) => updateAddress(address);

export const deleteAddress = async (id: number) => {
  getAddresses.clear();
  return makeApiCall<{ addressId: number }, void>(
    apiRoutes.deleteAddress, { body: { addressId: id } }
  );
};

export const setDefaultAddress = async (id: number) => {
  const address = await getAddressById(id);
  if (!address) throw new Error('An address with this id doesn\'t exist');

  await makeApiCall<{ addressId: number; isDefault: boolean }, APIAddressRequest>(apiRoutes.saveAddress, {
    body: {
      addressId: id,
      isDefault: true
    }
  });
  getAddresses.clear();

  await updateUser({
    lat: address.latitude,
    lng: address.longitude,
    pincode: address.pincode,
    storeId: address.storeId,
    storeName: address.storeName
  });

  return address;
};

export const getDefaultAddress = async () => {
  const addresses = await getAddresses();
  const defaultAddress = addresses.find(address => address.isDefault);
  if (defaultAddress) return defaultAddress;

  if (!addresses.length) return null;

  await setDefaultAddress(addresses[0].id);
  return addresses[0];
};

export const updateAddressWithKirana = async (addressId: number, { storeId, storeName }: { storeId: number; storeName: string }) => {
  const address = await getAddressById(addressId);

  if (!address) throw new Error('Couldn\'t find an address with this id');

  const updatedAddress = await updateAddress({ ...address, storeId, storeName });

  if (address.isDefault) {
    await updateUser({ storeId, storeName });
  }

  return updatedAddress;
};

export const deleteKiranaFromAddress = async (addressId: number) => updateAddressWithKirana(addressId, { storeId: 0, storeName: '' });

export const getStoreIdFromAddressId = async (addressId: number) => {
  const address = await getAddressById(addressId);

  if (!address) return null;

  return address.storeId || address.defaultStore || null;
};

export const getPincodeFromLatLng = async ({ lat, lng }: {lat: number; lng: number}): Promise<string> => {
  const geocoder = new window.google.maps.Geocoder();
  const pincode = new Promise<string>((resolve, reject) => {
    geocoder.geocode({ location: new window.google.maps.LatLng(lat, lng) }, (results, status) => {
      if (status === window.google.maps.GeocoderStatus.OK) {
        const len = results[0].address_components.length;
        resolve(results[0].address_components[len - 1].long_name);
      } else {
        reject();
      }
    });
  });
  return pincode;
};
