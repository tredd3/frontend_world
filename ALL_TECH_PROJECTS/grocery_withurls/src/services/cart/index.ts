import promiseMemoize from 'promise-memoize';
import { apiRoutes } from '../api-routes';
import makeApiCall from '../helpers/make-api-call';
import { APICart } from '../../types/api';
import { getUser } from '../user';
import { getCart as internalGetCart } from './internal';

export * from './pre-checkout';
export * from './post-checkout';
export * from './payment';

// PS: This method is only for use in old components that haven't been changed
// to use pivot data and are passing the data as such to multiple such components.
// If it is just a single component consuming the same data, refactor the component
// to rely on pivoted data instead.
export const deprecatedGetCart = promiseMemoize(async (storeId?: number): Promise<APICart> => {
  const { userId, preferences: { storeId: defaultStoreId } } = await getUser();
  const selectedStoreId = storeId || defaultStoreId;

  return makeApiCall<{}, APICart>(apiRoutes.getCart, {
    body: {
      userId,
      ...(selectedStoreId ? { storeId: selectedStoreId } : {})
    },
    ...(selectedStoreId ? { headers: { StoreId: `${selectedStoreId}` } } : {})
  });
}, { resolve: 'json' });

export const clearGetCartMemoization = () => internalGetCart.clear();
