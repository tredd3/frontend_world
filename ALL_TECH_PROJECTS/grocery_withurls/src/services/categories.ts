import promiseMemoize from 'promise-memoize';
import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { getUser } from './user';
import { CategoryPivot } from '../types';

type APICategory = {
  CategoryName: string;
  CategoryId: number;
  CategoryIcon: string;
  CategoryImage: string;
};

type APIRequest = {
  CategoryId?: number;
  Level: 2;
  StoreId: number;
  PageNo: number;
};

type APIResponse = {
  Categories: APICategory[];
};

const convertAPICategoryToCategory = (category: APICategory): CategoryPivot => ({
  id: category.CategoryId,
  name: category.CategoryName,
  icon: category.CategoryIcon,
  image: category.CategoryImage
});

export const getCategories = promiseMemoize(async (id: number) => {
  const { preferences: { storeId } } = await getUser();
  const { Categories } = await makeApiCall<APIRequest, APIResponse>(apiRoutes.getCategories, {
    body: {
      Level: 2,
      ...(id && { CategoryId: id }),
      StoreId: storeId || 0,
      PageNo: 0
    }
  });

  return Categories.map(convertAPICategoryToCategory);
}, { resolve: 'json' });
