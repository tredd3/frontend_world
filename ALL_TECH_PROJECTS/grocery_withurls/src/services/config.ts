import promiseMemoize from 'promise-memoize';
import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { getUser } from './user';
import { APIConfig, ConfigPivot } from '../types';
import { pivotConfig } from './helpers/pivots';

export const getConfig = promiseMemoize(async (): Promise<ConfigPivot> => {
  const { userId } = await getUser();
  const { Configurations: apiConfig } = await makeApiCall<
    { customerId: number },
    {Configurations: APIConfig}
  >(apiRoutes.config, { body: { customerId: userId } });

  return pivotConfig(apiConfig);
});

export const deprecatedGetConfigDATA = async () => {
  try {
    const config = await getConfig();

    let LAST_ORDER_DETAILS = {};
    if (config.lastOrderDetails && Object.keys(config.lastOrderDetails).length !== 0) {
      LAST_ORDER_DETAILS = {
        TOTAL_ITEMS: config.lastOrderDetails.totalItems,
        ORDER_ID: config.lastOrderDetails.orderId,
        STORE_NAME: config.lastOrderDetails.storeName,
        TOTAL_SHIPMENT: config.lastOrderDetails.totalShipment,
        DELIVERED_DATE: config.lastOrderDetails.deliveredDate,
        STORE_ID: config.lastOrderDetails.storeName,
        ORDER_AMOUNT: config.lastOrderDetails.orderAmount,
        DELIVERED_TIME: config.lastOrderDetails.deliveredDate
      };
    }

    return {
      DEFAULT_PAGE_SIZE: config.pageSize,
      QTY_ERROR_MESSAGE: config.quantityErrorMessage,
      INVALID_OTP: config.invalidOTPMessage,
      MAX_COD_LIMIT: config.maxCODLimit,
      DELIVERY_CHARGE_THRESHOLD_AMOUNT: config.deliveryChargeThreshold,
      UNREAD_NOTIFICATION_COUNT: config.unreadNotificationCount,
      DELIVERY_CHARGES: config.deliveryCharges,
      COD_LIMIT_ERROR_MESSAGE: config.codLimitErrorMessage,
      WELCOME_PRODUCT_SKU: config.welcomeProductSKU,
      GENERIC_ERROR_MESSAGE: config.genericErrorMessage,
      MAX_QTY_LIMIT_PER_ITEM: config.maxQuantityLimitPerItem,
      LAST_ORDER_DETAILS
    };
  } catch (e) {
    return {};
  }
};
