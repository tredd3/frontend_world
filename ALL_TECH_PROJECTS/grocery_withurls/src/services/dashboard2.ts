import promiseMemoize from 'promise-memoize';
import { APIDashboardProduct } from '../types/api';
import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { DashboardProductPivot } from '../types';
import { pivotDashboardProduct } from './helpers/pivots';
import { getUser } from './user';

type Banner = { id: number; imageUrl: string };
type BannerWithTitle = Banner & { title: string };
type WidgetOfType<T> = { type: T; subType: number };
type BannerWidget<T> = WidgetOfType<T> & { banners: Banner[] };

export type HorizontalScrollableIcons = WidgetOfType<1> & { banners: BannerWithTitle[] };
export type Pager = BannerWidget<3>;
export type ThinCarouselBanner = BannerWidget<4>;
export type BigSquareCarouselBanner = BannerWidget<5>;
export type SmallSquareCarouselBanner = BannerWidget<6> & { title: string }
export type HorizontalProducts = WidgetOfType<7> & {
  title: string;
  viewAllLabel: string;
  products: DashboardProductPivot[];
};
export type W4x = WidgetOfType<11> & { banners: BannerWithTitle[] };
export type W9x = WidgetOfType<12> & { banners: BannerWithTitle[] } & { title: string };
export type W4xWithBanner = BannerWidget<15> & { title: string };
export type W9xWithBanner = BannerWidget<16> & { title: string };

export type Widget =
  HorizontalScrollableIcons
  | Pager
  | ThinCarouselBanner
  | BigSquareCarouselBanner
  | SmallSquareCarouselBanner
  | HorizontalProducts
  | W4x
  | W9x
  | W4xWithBanner
  | W9xWithBanner

type DashboardRequestType = {
  pincode?: number;
  storeId?: number;
};

type APIBanner = {
  bannerId: number;
  imageUrl: string;
};

type APIMiscellaneous = {
  name: string;
  mscId: number;
  imageUrl: string;
}

type APIWidget = {
  type: Widget['type'];
  subType: number;
  title: string;
  deepLinkText: string;
  data: {
    banners: null | APIBanner[];
    products: null | APIDashboardProduct[];
    miscellaneous: null | APIMiscellaneous[];
  };
};

export type DashboardResponseType = {
  widgets: APIWidget[];
};

const widgetBase = (widget: APIWidget) => ({
  type: widget.type,
  subType: widget.subType
});

const apiBannerToBanner = (banner: APIBanner | APIMiscellaneous): Banner => ({
  id: (banner as APIBanner).bannerId || (banner as APIMiscellaneous).mscId,
  imageUrl: banner.imageUrl
});

const createBanners = ({ data: widgetData }: APIWidget): { banners: Banner[] } => ({
  banners: ((widgetData.banners || widgetData.miscellaneous || []) as (APIBanner | APIMiscellaneous)[])
    .map(apiBannerToBanner)
});

const pivotBanner = (widget: APIWidget) => ({
  ...widgetBase(widget),
  ...createBanners(widget)
});

const createBannersWithTitle = (widget: APIWidget): { banners: BannerWithTitle[] } => ({
  banners: ((widget.data.miscellaneous || [])as APIMiscellaneous[])
    .map(banner => ({
      ...apiBannerToBanner(banner),
      title: banner.name
    }))
});

const createProducts = (widget: APIWidget): { products: DashboardProductPivot[] } => ({
  products: widget.data.products!.map(pivotDashboardProduct)
});

const pivotBannersWithTitle = (widget: APIWidget) => ({
  ...widgetBase(widget),
  ...createBannersWithTitle(widget)
});

const pivotProducts = (widget: APIWidget) => ({
  ...widgetBase(widget),
  ...createProducts(widget)
});

export const pivotDashboardResponse = ({ widgets }: DashboardResponseType) => (
  widgets
    .filter(w => w.data.banners || w.data.miscellaneous || w.data.products)
    .map(widget => {
      switch (widget.type) {
        case 1: return pivotBannersWithTitle(widget) as HorizontalScrollableIcons;
        case 3: return pivotBanner(widget) as Pager;
        case 4: return pivotBanner(widget) as ThinCarouselBanner;
        case 5: return pivotBanner(widget) as BigSquareCarouselBanner;
        case 6:
          return {
            ...widgetBase(widget),
            title: widget.title,
            ...createBanners(widget)
          } as SmallSquareCarouselBanner;
        case 7:
          return {
            title: widget.title,
            viewAllLabel: widget.deepLinkText,
            ...pivotProducts(widget)
          } as HorizontalProducts;
        case 11: return pivotBannersWithTitle(widget) as W4x;
        case 12:
          return {
            title: widget.title,
            ...pivotBannersWithTitle(widget)
          } as W9x;
        case 15: return {
          ...widgetBase(widget),
          title: widget.title,
          ...createBanners(widget)
        } as W4xWithBanner;
        case 16: return {
          ...widgetBase(widget),
          title: widget.title,
          ...createBanners(widget)
        } as W9xWithBanner;
        default: return null;
      }
    })
    .filter(<T>(x: T | null): x is T => x !== null)
);

export const getWidgetDefinitions = promiseMemoize(async () => {
  const { preferences: { pincode, storeId = 0 } } = await getUser();

  const apiWidgets = await makeApiCall<DashboardRequestType, DashboardResponseType>(apiRoutes.getDashboard, {
    body: { pincode, storeId }
  });

  return pivotDashboardResponse(apiWidgets);
}, { resolve: 'json' });
