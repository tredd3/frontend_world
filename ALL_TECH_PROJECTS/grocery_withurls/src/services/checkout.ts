import promiseMemoize from 'promise-memoize';
import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { getUser } from './user';
import { getAddressById } from './address';
import { getCart } from './cart';
import { APIShipment } from '../types/api';

type CheckoutRequestBody = {
  deliveryDate: string;
  timeslot: string;
  storeId: number | undefined;
  cartId: number;
  userId: number;
  userMobile: number;
  customerAddressId: number | undefined;
  isAutoSelected: number;
}
export type CheckoutRequestData = {
  deliveryDate: string;
  timeslot: string;
  addressId: number;
  storeId: number;
 }

export type CheckoutData = {
  CartId: string;
  UserId: string;
  Shipments: APIShipment[];
  TotalAmount: number;
  TotalMRPDiscount: number;
  TotalOrderLevelDiscount: number;
  TotalDeliveryCharge: number;
  TotalItem: number;
  TotalCouponDiscount: number;
  TotalPay: number;
  ConfiguredDeliveryCharge: string;
}

export const checkoutOrder = promiseMemoize(async (payload: CheckoutRequestData) => {
  const user = await getUser();
  const address = await getAddressById(Number(payload.addressId));
  const cart = await getCart();

  const { storeId } = payload;

  const requestPayload: CheckoutRequestBody = {
    storeId,
    cartId: cart.cartId,
    userId: user.userId,
    userMobile: user.phoneNumber,
    customerAddressId: address && address.id,
    isAutoSelected: 0,
    deliveryDate: payload.deliveryDate,
    timeslot: payload.timeslot
  };

  const customHeader = { StoreId: String(storeId) };

  const data = await makeApiCall<CheckoutRequestBody, CheckoutData>(apiRoutes.checkout,
    {
      headers: customHeader,
      body: {
        ...requestPayload
      }
    });
  return data;
}, {
  resolve: 'json'
});
