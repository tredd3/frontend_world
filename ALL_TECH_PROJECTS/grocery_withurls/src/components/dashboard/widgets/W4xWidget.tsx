
/** @jsx jsx */
import { jsx, css } from '@emotion/core';
import { Link } from 'react-router-dom';
import Image from '../../uiControls/Image';
import { W4x } from '../../../services/dashboard2';
import linkStyles from '../linkStyles';
import { categoryUrl } from '../../../helpers/urls';

const CONTAINER_PADDING = 10;
const CELL_PADDING = 15;
const IMAGE_WIDTH = (window.outerWidth / 2) - (2 * CONTAINER_PADDING) - (2 * CELL_PADDING);

const styles = {
  wrapper: css`
    padding: ${CONTAINER_PADDING}px;
    border-bottom: 5px solid #F2F2F2;
  `,
  header: css`
    display: flex;
    justify-content: space-between;
    align-items: center;
    font-size: 16px;
  `,
  grid: css`
    display: grid;
    grid-template-columns: 1fr 1fr;
    margin-top: 10px;
    grid-gap: 1px 1px;
    background-color: #F2F2F2;
  `,
  gridCell: css`
    text-align: center;
    background-color: #fff;
    padding: ${CELL_PADDING}px;
    display: flex;
    flex-direction: column;
  `
};

type Props = {
  widgetData: W4x;
}

const W4xWidget = ({
  widgetData: {
    banners
  }
}: Props) => {
  if (banners.length) {
    return (
      <section css={styles.wrapper}>
        <div css={styles.grid}>
          {banners.map(({ imageUrl, title, id }, index) => (
            <Link
              data-testid={`banner-link-${index}`}
              key={id}
              css={[linkStyles, styles.gridCell]}
              to={categoryUrl(id)}
            >
              <div className="imageContainer">
                <Image width={IMAGE_WIDTH} alt="img" src={imageUrl} />
              </div>
              <div css={css`font-size: 13px;`}>{title}</div>
            </Link>
          ))}
        </div>
      </section>
    );
  }

  return null;
};

export default W4xWidget;
