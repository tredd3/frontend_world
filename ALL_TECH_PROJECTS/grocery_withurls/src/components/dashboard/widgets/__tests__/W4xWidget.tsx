import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import W4xWidget from '../W4xWidget';
import { pivotDashboardResponse, W4x, DashboardResponseType } from '../../../../services/dashboard2';
import mockDB from '../../../../mocks/db.json';

jest.mock('../../../../helpers/analytics');
jest.mock('../../../../intents');

const mockWidgetData = pivotDashboardResponse(mockDB.dashboard.Data as DashboardResponseType)
  .find(({ type }) => type === 11) as W4x;

describe('W4xWidget widget tests', () => {
  it('should render the widget correctly', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <W4xWidget
          widgetData={mockWidgetData}
        />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
