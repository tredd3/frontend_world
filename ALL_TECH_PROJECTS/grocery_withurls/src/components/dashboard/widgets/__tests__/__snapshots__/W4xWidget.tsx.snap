// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`W4xWidget widget tests should render the widget correctly 1`] = `
<DocumentFragment>
  .emotion-9 {
  padding: 10px;
  border-bottom: 5px solid #F2F2F2;
}

.emotion-8 {
  display: grid;
  grid-template-columns: 1fr 1fr;
  margin-top: 10px;
  grid-gap: 1px 1px;
  background-color: #F2F2F2;
}

.emotion-1 {
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: none;
  text-align: center;
  background-color: #fff;
  padding: 15px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-1:hover,
.emotion-1:focus,
.emotion-1:visited,
.emotion-1:active {
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: none;
}

.emotion-0 {
  font-size: 13px;
}

<section
    class="emotion-9"
  >
    <div
      class="emotion-8"
    >
      <a
        class="emotion-1"
        data-testid="banner-link-0"
        href="/categories/10003400"
      >
        <div
          class="imageContainer"
        >
          <span
            class=" lazy-load-image-background "
            style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; width: 462px;"
          >
            <img
              alt="img"
              src="https://preprod-static.jiomoney.com/static//delivery_associate/DELASSO1568198485851000001.png?impolicy=resizeforwidth&w=462"
              width="462"
            />
          </span>
        </div>
        <div
          class="emotion-0"
        >
          Mom & Baby
        </div>
      </a>
      <a
        class="emotion-1"
        data-testid="banner-link-1"
        href="/categories/10003388"
      >
        <div
          class="imageContainer"
        >
          <span
            class=" lazy-load-image-background "
            style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; width: 462px;"
          >
            <img
              alt="img"
              src="https://preprod-static.jiomoney.com/static//delivery_associate/DELASSO1568198494531000001.png?impolicy=resizeforwidth&w=462"
              width="462"
            />
          </span>
        </div>
        <div
          class="emotion-0"
        >
          Beauty & Hygiene
        </div>
      </a>
      <a
        class="emotion-1"
        data-testid="banner-link-2"
        href="/categories/10003627"
      >
        <div
          class="imageContainer"
        >
          <span
            class=" lazy-load-image-background "
            style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; width: 462px;"
          >
            <img
              alt="img"
              src="https://preprod-static.jiomoney.com/static//delivery_associate/DELASSO1568198499921000001.png?impolicy=resizeforwidth&w=462"
              width="462"
            />
          </span>
        </div>
        <div
          class="emotion-0"
        >
          Foodgrains, Oils And Masala
        </div>
      </a>
      <a
        class="emotion-1"
        data-testid="banner-link-3"
        href="/categories/10003397"
      >
        <div
          class="imageContainer"
        >
          <span
            class=" lazy-load-image-background "
            style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; width: 462px;"
          >
            <img
              alt="img"
              src="https://preprod-static.jiomoney.com/static//delivery_associate/DELASSO1568198503111000001.png?impolicy=resizeforwidth&w=462"
              width="462"
            />
          </span>
        </div>
        <div
          class="emotion-0"
        >
          Health Drinks & Beverages
        </div>
      </a>
    </div>
  </section>
</DocumentFragment>
`;
